lib/prelude.dx

```diff
@@ -348,6 +355,39 @@ instance [Ord a, Ord b] Ord (a & b)
   (>) = \(x1,x2) (y1,y2). x1 > y1 || (x1 == y1 && x2 > y2)
   (<) = \(x1,x2) (y1,y2). x1 < y1 || (x1 == y1 && x2 < y2)
 
+instance Eq Ordering
+  (==) = \x y. OToW8 x == OToW8 y
+
+interface Monoid a
+  mempty : a
+  mcombine : a -> a -> a  -- can't use `<>` just for parser reasons?
+
+def (<>) [Monoid a] : a -> a -> a = mcombine
+
+def scan (init:a) (body:n->a->(a&b)) : (a & n=>b) =
+  swap $ runState init \s. for i.
+    c = get s
+    (c', y) = body i c
+    s := c'
+    y
+
+def fold (init:a) (body:(n->a->a)) : a = fst $ scan init \i x. (body i x, ())
+
+def compare [Ord a] (x:a) (y:a) : Ordering =
+  if x < y
+    then LT
+    else if x == y
+      then EQ
+      else GT
+
+instance Monoid Ordering
+  mempty = EQ
+  mcombine = \x y.
+    case x of
+      LT -> LT
+      GT -> GT
+      EQ -> y
```

Comment:
Only parameter `x` is returned, parameter `y` is ignored. This seems not robust against e.g. ordering for unstable sorts.

Is this information-discarding standard and actually okay? Reminds me of the `Alternative` typeclass.

Reply:
I'm actually not sure what you mean, could you clarify? I don't see how `x` is returned by `mcombine`, and y is only ignored if `x` is not `EQ`, right?

I definitely don't know enough Haskell to know what is standard. But, this sentence from the [Haskell wiki article for monoid](https://wiki.haskell.org/Monoid) is suggestive: "First, `Ordering`, the standard type which Haskell uses for the result of `compare` functions, has a "lexicographic" combination operation, where `mappend` essentially takes the first non-equality result." (although I couldn't find its actual monoid instance implementation).

Reply:
Thanks for the clarification! My mistake, I am not familiar with the standard `Monoid Ordering` definition - I wonder if we find a Haskell reference implementation?

My question was basically:
* How does `LT <> GT == LT` make sense?
* Same for other variants: `GT <> EQ == GT`, etc.

Reply:
[Here](https://hackage.haskell.org/package/base-4.14.1.0/docs/src/GHC.Base.html#line-406) is the reference implementation from `base`.

Reply:
Thanks Shaurya! Looks like this is the standard implementation, pretty non-controversial.

Reply:
Thanks sureyeaah. Because I want to use this type for lexicographic sorting, I think we need `LT <> GT == LT`, and in that context it does make sense. Similarly it's not commutative because flipping one of the tables you're comparing should yield a different lexicographic ordering. But it looks like I'm mimicking what Haskell itself is doing, so I think this is necessary for lexicographic ordering.